src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Repository/ParquetDataRepository.cs

```diff
@@ -0,0 +1,161 @@
+
+        var command = duckDbConnection.SqlBuilder(
+            $"""
+             SELECT count(*)
+             FROM '{parquetPathResolver.DataPath(dataSetVersion):raw}'
+             """
+        );
+
+        command.AppendIf(!where.IsEmpty(), $"WHERE {where}");
+
+        return await command.QuerySingleAsync<long>(cancellationToken: cancellationToken);
+    }
+
+    public async Task<IEnumerable<IDictionary<string, object?>>> ListRows(
+        DataSetVersion dataSetVersion,
+        IEnumerable<string> columns,
+        IInterpolatedSql where,
+        IEnumerable<Sort>? sorts = null,
+        int page = 1,
+        int pageSize = 1000,
+        CancellationToken cancellationToken = default)
+    {
+        using var timing = MiniProfiler.Current
+            .Step($"{nameof(ParquetDataRepository)}.{nameof(ListRows)}");
+
+        var dataPath = parquetPathResolver.DataPath(dataSetVersion);
+
+        var whereFragment = new DuckDbSqlBuilder()
+            .AppendIf(!where.IsEmpty(), $"WHERE {where}");
+
+        var orderings = (sorts ?? [])
+            .Select(s => $"{s.Field} {s.Direction.ToString().ToUpper()}")
+            .ToList();
+
+        var orderByFragment = new DuckDbSqlBuilder()
+            .AppendIf(orderings.Count != 0, $"ORDER BY")
+            .AppendRange(orderings, joinString: ",\n");
+
+        var pageOffset = (page - 1) * pageSize;
+
+        // We essentially split this query into two sub-queries:
+        //
+        // 1. The main query which is offset paginated and gathers the row ids
+        // 2. Another query to fetch the rows using the ids from the main query (i.e. a 'deferred' join)
+        //
+        // This 'deferred join' technique is more efficient than a single query and helps to reduce
+        // the performance penalty of using offset pagination having to scan through many rows.
+        var command = duckDbConnection.SqlBuilder(
+            $"""
+             WITH {DataIdsAlias:raw} AS (
+                SELECT {DataTable.Ref().Id:raw}
+                FROM '{dataPath:raw}' AS {DataTable.TableName:raw}
+                {whereFragment}
+                {orderByFragment}
+                LIMIT {pageSize}
+                OFFSET {pageOffset}
+             )
+             SELECT {columns.Select(DataTable.Ref().Col).JoinToString(",\n"):raw}
+             FROM '{dataPath:raw}' AS {DataTable.TableName:raw}
+             JOIN {DataIdsAlias:raw} ON {DataIdsAlias:raw}.id = {DataTable.Ref().Id:raw}
+             {orderByFragment}
+             """
+        );
+
+        return (await command.QueryAsync(cancellationToken: cancellationToken))
+            .Cast<IDictionary<string, object?>>();
+    }
+
+    public async Task<ISet<string>> ListColumns(
+        DataSetVersion dataSetVersion,
+        CancellationToken cancellationToken = default)
+    {
+        var command = duckDbConnection.SqlBuilder(
+            $"DESCRIBE SELECT * FROM '{parquetPathResolver.DataPath(dataSetVersion):raw}' LIMIT 1");
```

Comment:
`LIMIT 1` just means that only one row is returned, rather than all rows. We still get a complete set of column information.

I'm not entirely sure if adding the `LIMIT 1` really makes a difference as we're using `DESCRIBE`. I'm not sure what happens under the hood with DuckDB, but I'd hope it doesn't actually try and retrieve all rows when using `DESCRIBE SELECT *`, but in-case it does, I've added in the `LIMIT 1` as a precaution.

Reply:
Sorry, I misinterpreted the syntax. I thought it was going to describe all the columns (each of which would be a row of descriptive data) - and then take only one of those resulting rows. Which would end up just being the description for 1 column and ignoring the others!